File: coding/src/com/algos/li10_tree/bst/revised/L333_LargestBstSubtree.java
package  com.algos.li10_tree.bst.revised;

import  com.algos.li30_model.TreeNode;


public class L333_LargestBstSubtree {
    
	public static void main(String[] args) {
		L333_LargestBstSubtree obj = new L333_LargestBstSubtree();
		
        TreeNode root = new TreeNode(6);
		root.left = new TreeNode(3);
		root.right = new TreeNode(5);
        root.left.right = new TreeNode(2);
		root.right.left = new TreeNode(0);
        root.left.right.right = new TreeNode(1);
		
        System.out.println(obj.maxSumBST(root));
    }
	
    // Problem: https://leetcode.com/problems/largest-bst-subtree/
	// https://www.lintcode.com/problem/910/
    // Idea: Use a data structure int[]{size, min, max}
	// time: O(n)
    // space: O(1)
	private int MAX = Integer.MAX_VALUE, MIN = Integer.MIN_VALUE;
	public int maxSumBST(TreeNode root) {
		return maxSizeBstRec(root).size;
	}

	private NodeState maxSizeBstRec(TreeNode root) {
		if(root == null)	// a empty tree is bst of size 0
			return new NodeState(0, MAX, MIN);
		
		NodeState left = maxSizeBstRec(root.left);
		NodeState right = maxSizeBstRec(root.right);
		
		if(root.val > left.max && root.val < right.min){
			return new NodeState(left.size+right.size+1, 
				Math.min(left.min, root.val), Math.max(right.max, root.val)); // Math.min to avoid null node max value. Otherwise, it will be left.min
		}else{
			return new NodeState(Math.max(left.size, right.size), 
				MIN, MAX);
		}
	}
}
class NodeState{
	public int size, min, max;
	public NodeState(int size, int min, int max) {
		this.size = size;
		this.min = min;
		this.max = max;
	}
}

File: dsa/concepts/10_trees/bst/python/l333__largest_bst_subtree.py
# Tree / DFS
# Problem: https://www.lintcode.com/problem/910/
# Idea:
# 1. Use a helper structure NodeState (size, min, max) for each subtree.
# 2. If current node satisfies BST property w.r.t left and right subtree, update size, min, max.
# 3. Otherwise, propagate the largest size found so far but invalidate min/max for BST check.
# Time: O(n)
# Space: O(1) (ignoring recursion stack)

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class NodeState:
    def __init__(self, size, min_val, max_val):
        self.size = size
        self.min = min_val
        self.max = max_val

MAX_INT = float('inf')
MIN_INT = float('-inf')

def maxSumBST(root):
    def maxSizeBstRec(node):
        if not node:
            return NodeState(0, MAX_INT, MIN_INT)  # empty tree is BST of size 0
        
        left = maxSizeBstRec(node.left)
        right = maxSizeBstRec(node.right)
        
        if node.val > left.max and node.val < right.min:
            size = left.size + right.size + 1
            min_val = min(left.min, node.val)
            max_val = max(right.max, node.val)
            return NodeState(size, min_val, max_val)
        else:
            return NodeState(max(left.size, right.size), MIN_INT, MAX_INT)
    
    return maxSizeBstRec(root).size


# -------- Test --------
root = TreeNode(6)
root.left = TreeNode(3)
root.right = TreeNode(5)
root.left.right = TreeNode(2)
root.right.left = TreeNode(0)
root.left.right.right = TreeNode(1)

print(maxSumBST(root))  # Expected: size of largest BST subtree
